File: youkai-front/src/domain/youkai/checkdigit/state/AttackCharacters.ts
import A31F from "./A31F";
import CodeToCharacterConverter from "@/domain/youkai/checkdigit/converter/CodeToCharacterConverter";

export default class AttackCharacters {
    constructor(
        private readonly converter: CodeToCharacterConverter,
        private readonly charCodes: number[]
    ) { }

    public static readonly MIN_CHARS_LENGTH = 3;
    public static readonly MAX_CHARS_LENGTH = 14;

    public getOf(index: number): number {
        return this.charCodes[index];
    }

    public charLength(): number {
        return this.charCodes.length;
    }

    public increment(): AttackCharacters {
        const converter = this.converter;
        const newCodes = this.charCodes.slice();
        for (let i = 0; i < newCodes.length; i++) {
            const before = newCodes[i];
            newCodes[i] = converter.incrementCode(before);
            if (newCodes[i] > before) break; // 繰り上がりなし
        }
        return new AttackCharacters(converter, newCodes);
    }

    public fixInvalid(): AttackCharacters {
        if (!this.isInvalid()) return this;

        const converter = this.converter;
        const newCodes = this.charCodes.slice();
        for (let i = 0; i < newCodes.length; i++) {
            if (!converter.isInvalidCharCode(newCodes[i])) continue;
            newCodes[i] = converter.incrementCode(newCodes[i]);
        }
        return new AttackCharacters(converter, newCodes);
    }

    public isInvalid(): boolean {
        return this.charCodes
            .some(code => this.converter.isInvalidCharCode(code));
    }

    public dumpHexText(): string {
        return this.charCodes
            .map(code => A31F.intToHex(code))
            .join(" ")
            .trim();
    }

    public toString(): string {
        return this.charCodes
            .map(code => this.converter.convert(code))
            .join("")
            .trim();
    }

    public equals(o: AttackCharacters): boolean {
        if (this === o) return true;
        if (o === null) return false;
        return this.charCodes.toString() === o.charCodes.toString();
    }

    public minimum(): AttackCharacters {
        return AttackCharacters.initialize(this.charCodes.length, this.converter);
    }

    public static initialize(charCount: number, converter = new CodeToCharacterConverter()): AttackCharacters {
        const initialCode = converter.minCode();
        const values = Array(charCount).fill(initialCode);
        return new AttackCharacters(converter, values);
    }

    public static withText(passwordText: string, converter = new CodeToCharacterConverter()): AttackCharacters {
        const codes = passwordText.split("")
            .map(oneCher => converter.reverceConvert(oneCher));
        return new AttackCharacters(converter, codes);
    }
}

File: youkai-console/src/main/java/com/github/kazuhitom/youkai/console/checkdigit/state/AttackCharacters.java
package com.github.kazuhitom.youkai.console.checkdigit.state;

import com.github.kazuhitom.youkai.console.checkdigit.converter.CodeToCharacterConverter;

import java.util.Arrays;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.joining;

public class AttackCharacters {
    private final int[] charCodes;
    private final CodeToCharacterConverter converter;

    public int getOf(int index) {
        return charCodes[index];
    }

    public int charLength() {
        return charCodes.length;
    }

    public AttackCharacters increment() {
        int[] newCodes = Arrays.copyOf(charCodes, charCodes.length);
        for (int i = 0; i < newCodes.length; i++) {
            int before = newCodes[i];
            newCodes[i] = converter.incrementCode(before);
            if (newCodes[i] > before) break; // 繰り上がりなし
        }
        return new AttackCharacters(converter, newCodes);
    }

    public AttackCharacters fixInvalid() {
        if (!isInvalid()) return this;

        int[] newCodes = Arrays.copyOf(charCodes, charCodes.length);
        for (int i = 0; i < newCodes.length; i++) {
            if (!converter.isInvalidCharCode(newCodes[i])) continue;
            newCodes[i] = converter.incrementCode(newCodes[i]);
        }
        return new AttackCharacters(converter, newCodes);
    }

    public boolean isInvalid() {
        return Arrays.stream(charCodes)
                .anyMatch(converter::isInvalidCharCode);
    }

    public String dumpHexText() {
        return Arrays.stream(charCodes)
                .mapToObj(code -> String.format("%02X", code))
                .collect(joining(" "));
    }

    @Override
    public String toString() {
        return Arrays.stream(charCodes)
                .mapToObj(converter::convert)
                .map(c -> c.toString())
                .collect(joining());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AttackCharacters that = (AttackCharacters) o;
        return Arrays.equals(charCodes, that.charCodes);
    }

    @Override
    public int hashCode() {
        return Arrays.hashCode(charCodes);
    }

    public AttackCharacters minimum() {
        return initialize(charCodes.length, converter);
    }

    public static AttackCharacters initialize(int charCount) {
        return new AttackCharacters(new int[charCount]);
    }

    public static AttackCharacters initialize(int charCount, CodeToCharacterConverter converter) {
        int initialCode = converter.minCode();
        int[] values = IntStream.range(0, charCount)
                .map(i -> initialCode)
                .toArray();
        return new AttackCharacters(converter, values);
    }


    public static AttackCharacters withText(String passwordText) {
        return withText(passwordText, new CodeToCharacterConverter());
    }

    public static AttackCharacters withText(String passwordText, CodeToCharacterConverter converter) {
        int[] codes = passwordText.chars()
                .map(oneCher -> converter.reverceConvert((char) oneCher))
                .toArray();
        return new AttackCharacters(converter, codes);
    }

    public AttackCharacters(int... charCodes) {
        this(new CodeToCharacterConverter(), charCodes);
    }

    public AttackCharacters(CodeToCharacterConverter converter, int... charCodes) {
        this.converter = converter;
        this.charCodes = charCodes;
    }
}
